storage/src/main/java/org/apache/kafka/storage/internals/log/RemoteIndexCache.java

```diff
@@ -0,0 +1,657 @@
+        }
+
+        public TransactionIndex txnIndex() {
+            return txnIndex;
+        }
+
+        // Visible for testing
+        public boolean isCleanStarted() {
+            return cleanStarted;
+        }
+
+        // Visible for testing
+        public boolean isMarkedForCleanup() {
+            return markedForCleanup;
+        }
+
+        public OffsetPosition lookupOffset(long targetOffset) {
+            lock.readLock().lock();
+            try {
+                if (markedForCleanup) throw new IllegalStateException("This entry is marked for cleanup");
+                else return offsetIndex.lookup(targetOffset);
+            } finally {
+                lock.readLock().unlock();
+            }
+        }
+
+        public OffsetPosition lookupTimestamp(long timestamp, long startingOffset) throws IOException {
+            lock.readLock().lock();
+            try {
+                if (markedForCleanup) throw new IllegalStateException("This entry is marked for cleanup");
+
+                TimestampOffset timestampOffset = timeIndex.lookup(timestamp);
+                return offsetIndex.lookup(Math.max(startingOffset, timestampOffset.offset));
+            } finally {
+                lock.readLock().unlock();
+            }
+        }
+
+        public void markForCleanup() throws IOException {
+            lock.writeLock().lock();
+            try {
+                if (!markedForCleanup) {
+                    markedForCleanup = true;
+                    offsetIndex.renameTo(new File(Utils.replaceSuffix(offsetIndex.file().getPath(), "", LogFileUtils.DELETED_FILE_SUFFIX)));
+                    timeIndex.renameTo(new File(Utils.replaceSuffix(timeIndex.file().getPath(), "", LogFileUtils.DELETED_FILE_SUFFIX)));
+                    txnIndex.renameTo(new File(Utils.replaceSuffix(txnIndex.file().getPath(), "", LogFileUtils.DELETED_FILE_SUFFIX)));
+                }
+            } finally {
+                lock.writeLock().unlock();
+            }
+        }
+
+        public void cleanup() throws IOException {
+            markForCleanup();
+            // no-op if clean is done already
+            if (!cleanStarted) {
+                cleanStarted = true;
+
+                List<StorageAction<Void, Exception>> actions = Arrays.asList(() -> {
+                    offsetIndex.deleteIfExists();
+                    return null;
+                }, () -> {
+                    timeIndex.deleteIfExists();
+                    return null;
+                }, () -> {
+                    txnIndex.deleteIfExists();
+                    return null;
+                });
+
+                tryAll(actions);
+            }
+        }
+
+        public void close() {
+            Utils.closeQuietly(offsetIndex, "OffsetIndex");
+            Utils.closeQuietly(timeIndex, "TimeIndex");
+            Utils.closeQuietly(txnIndex, "TransactionIndex");
+        }
+
+        @Override
+        public String toString() {
```

Comment:
This just prints the object reference, that is why the scala file uses the file name instead.